http/http.go

```diff
@@ -59,18 +61,19 @@ func (no noLocalTransport) RoundTrip(req *http.Request) (*http.Response, error)
 				no.errlog.WithError(err).Error("Cancelled request due to error in address parsing")
 				return
 			}
-			ip := net.ParseIP(host)
-			if ip == nil {
+
+			ips, err := net.LookupIP(host)
```

Comment:
Great work! I have one question though. Does this mean will have to resolve the same host IP twice per request? (one once we get the connection and another during the normal request flow)

Been looking though how `httptrace` works and it does have a `DNSDone` hook which we could potentially use to avoid the dual lookup. The DNSDone hook will only be called when using non-idle connections (which should be fine since we're starting with a clean slate after each deploy), but it's called before actually establishing new connections, which is what we want:

- [DNSDone called here](https://cs.opensource.google/go/go/+/master:src/net/lookup.go;l=328;drc=refs%2Ftags%2Fgo1.17.3;bpv=0;bpt=0) after a DNS query
- [dns query invoked at the beginning of this block, connection created at the end](https://cs.opensource.google/go/go/+/master:src/net/dial.go;l=406-432;drc=refs%2Ftags%2Fgo1.17.2;bpv=1;bpt=1?q=dial.go&ss=go%2Fgo)

There will still be some competition on the context cancelling vs connection creation after calling `cancel()`, but I think this solution should be good enough for what we want to achieve.

I used the following playground to get the stack trace for what happens behind the scenes: https://play.golang.org/p/k_XPwRcuvMZ

Reply:
I think you're right that we do two dns lookups per request with this change, but it's likely cached most of the time? 

I think this change makes sense to fix the immediate issue, but overall it seems that `httptrace` is not really the right approach for this (there's also this issue https://github.com/netlify/netlify-commons/issues/255). I think a better approach might be to set a custom `DialContext` on the `http.Transport`. Then you can close the connection if the remote addr is not allowed, or even resolve the hostname first and return an error before you attempt a connection.

Reply:
> I think you're right that we do two dns lookups per request with this change, but it's likely cached most of the time?

that depends on the OS distribution and configuration underneath, but yeah, it should be. Go itself doesn't seem to cache anything.

>  I think a better approach might be to set a custom DialContext on the http.Transport

I agree